ARCs/arc-0020.md

```diff
@@ -0,0 +1,485 @@
+and relevant specification details.
+
+### **1. IDE Support**
+
+There are two general ways of interacting with an Application. One is by making an
+application call transaction and the other is by reading global or local state.
+The state variables are named.
+
+An imaginary Application called SugarSupplierContract offers
+a price for refined sugar to the public. This can be read from the
+global state variable "SugarPrice." A retailer negotiates a discount
+wholesale price. This is offered to the customer via a local state variable
+called "YourPrice." The Application also offers a method to calculate 
+the estimated shipping price for a given volume of sugar.
+
+The retailer would like to be integrate with this SugarSupplierContract
+using the development tool of their choice.
+
+The retailer, in this scenario already has the Application Specification
+available, so they give this to their IDE and the following proxy class is
+generated. The below is just a C# example.
+
+```cs
+   public abstract class SugarSupplierContract : SmartContractReference
+    {
+        [Storage(StorageType.Global)]
+        public ulong SugarPrice;
+
+        [Storage(StorageType.Local)]
+        public ulong YourPrice;
+
+  
+        public abstract ulong ShippingCost(ulong VolumeInKilos);
+    } 
+```
+
+Proxies can support server or dapp (frontend) calls to an Algorand node
+or even contract to contract invocations for compilers.
+
+Even without the IDE to tooling support, a dapp author
+still needs to understand how the SugarSupplierContract is intended
+to be used, so information must be provided somehow.
+
+To achieve this goal, this ARC document proposes that the
+Application Specification include a single ARC-4 Contract
+description, and that the ARC-4 Contract section be updated to
+include locals and globals:
+
+```typescript
+interface Contract {
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
+  /**
+   * Optional object listing the contract instances across different networks
+   */
+  networks?: {
+    /**
+     * The key is the base64 genesis hash of the network, and the value contains
+     * information about the deployed contract in the network indicated by the
+     * key
+     */
+    [network: string]: {
+      /** The app ID of the deployed contract in this network */
+      appID: number;
+    }
+  }
+  /** All of the methods that the contract implements */
+  methods: Method[];
+
+  /** All local variables intended for consumption as part of this contract **/
+  locals: Variable[];
+
+  /** All global variables intended for consumption as part of this contract **/
+  globals: Variable[];
+
+}
+```
+
+There needs to be a _Variable_ specification too, following similar style as
+the _Method_ entity:
+
+```typescript
+interface Variable {
+  /** The name of the variable **/
+  name: string;
```

Comment:
Sorry I haven't looked at the proposal in details.
But one question is how do you encode `name`, especially for non-utf-8 names.
string is indeed limited to utf-8, but not keys.

Reply:
Hi there. Thanks. This Arc is about a metadata format that allows a contract to be expressed as a 'class' where methods (abi signatures) and fields (state vars) are named as utf8 . in both cases the underlying byte array can hold anything but having method and property names as anything other than utf8 strings is unlikely to be directly supported by HLLs. Yes the code generators implementing the proxy classes on the basis of these metadata files to call the contracts could add mapping code to write or read from byte array keys but this would, it seems to me, be superfluous complexity. Though am happy to be told why here non utf8 property names should be supported.

Reply:
I suspect a common use case would be using a key that is the raw 32 bytes of an address, or is a hash of something.

Looking ahead to boxes, I think those will be very common names for boxes.

Reply:
It will also be very reasonable for a HLL to compile a variable down to a single letter key, in order to make a full 127bytes available in the value. So perhaps there should be a `name` and a `key`.  The name is human readable and might be the variable name in the HLL, even if the Algorand state key is called something smaller and less understandable.  Perhaps `key` is defaulted to `name` to simplify a common case, but `key`, if specified, is specified as b64 encoded bytes.